onnxruntime/python/tools/transformers/models/gpt2/gpt2_helper.py

```diff
@@ -522,6 +522,11 @@ def optimize_onnx(
             # Potentially, we can add other optimization like unpad for effective transformer
             optimization_options.attention_mask_format = AttentionMaskFormat.MaskIndexEnd
 
+        # TODO(hasesh): Investigate parity issue for GPT-2 fp16 when SkipLayerNormalization
+        # is enabled
+        if is_float16:
+            optimization_options.enable_skip_layer_norm = False
```

Comment:
It seems like there are some parity issues with fp16 + SLN. I added the SLN op to the op_block_list like this - arguments.extend(["--op_block_list", "Add", "LayerNormalization", "SkipLayerNormalization", "FastGelu"]) which should in theory still make it the same as the graph with LayerNormalizations + Adds. Am I missing something obvious: @tlwu @wangyems ?

Reply:
The order of Add might matter. In the original onnx graph, the order of add is input + bias + skip, in your SLN kernel it is like input + skip + bias. When fp16 is used, the result might be different.